lib/Frontend/CompilerInvocation.cpp

```diff
@@ -223,52 +235,86 @@ class ArgsToFrontendInputsConverter {
     return false;
   }
 };
-} // namespace swift
+class FrontendArgsToOptionsConverter {
+private:
+  DiagnosticEngine &Diags;
+  const llvm::opt::ArgList &Args;
+  FrontendOptions &Opts;
 
-// This is a separate function so that it shows up in stack traces.
-LLVM_ATTRIBUTE_NOINLINE
-static void debugFailWithAssertion() {
-  // This assertion should always fail, per the user's request, and should
-  // not be converted to llvm_unreachable.
-  assert(0 && "This is an assertion!");
-}
+  Optional<const std::vector<std::string>>
+      cachedOutputFilenamesFromCommandLineOrFilelist;
 
-// This is a separate function so that it shows up in stack traces.
-LLVM_ATTRIBUTE_NOINLINE
-static void debugFailWithCrash() {
-  LLVM_BUILTIN_TRAP;
-}
+  void handleDebugCrashGroupArguments();
+
+  void computeDebugTimeOptions();
+  bool computeFallbackModuleName();
+  bool computeModuleName();
+  bool computeOutputFilenames();
+  void computeDumpScopeMapLocations();
+  void computeHelpOptions();
+  void computeImplicitImportModuleNames();
+  void computeImportObjCHeaderOptions();
+  void computeLLVMArgs();
+  void computePlaygroundOptions();
+  void computePrintStatsOptions();
+  void computeTBDOptions();
+
+  void setUnsignedIntegerArgument(options::ID optionID, unsigned max,
+                                  unsigned &valueToSet);
+
+  FrontendOptions::ActionType determineRequestedAction() const;
+
+  bool setUpForSILOrLLVM();
+
+  /// Determine the correct output filename when none was specified.
+  ///
+  /// Such an absence should only occur when invoking the frontend
+  /// without the driver,
+  /// because the driver will always pass -o with an appropriate filename
+  /// if output is required for the requested action.
+  bool deriveOutputFilenameFromInputFile();
+
+  /// Determine the correct output filename when a directory was specified.
+  ///
+  /// Such a specification should only occur when invoking the frontend
+  /// directly, because the driver will always pass -o with an appropriate
+  /// filename if output is required for the requested action.
+  bool deriveOutputFilenameForDirectory(StringRef outputDir);
 
+  std::string determineBaseNameOfOutput() const;
+
+  void determineSupplementaryOutputFilenames();
+
+  /// Returns the output filenames on the command line or in the output
+  /// filelist. If there
+  /// were neither -o's nor an output filelist, returns an empty vector.
+  const std::vector<std::string> &getOutputFilenamesFromCommandLineOrFilelist();
```

Comment:
Now that this has a simpler type, is there any reason not to use ArrayRef here?

Reply:
D'Oh! Done.